js/lib/BTCUtils.js

```diff
@@ -0,0 +1,659 @@
+   * @returns {number}      The number of non-flag bytes in the VarInt
+   */
+  determineVarIntDataLength: (flag) => {
+    if (flag == 0xff) {
+      return 8 ; // one-byte flag, 8 bytes data
+    }
+    if (flag == 0xfe) {
+      return 4;  // one-byte flag, 4 bytes data
+    }
+    if (flag == 0xfd) {
+      return 2;  // one-byte flag, 2 bytes data
+    }
+
+    return 0;  // flag is data
+  },
+
+  /**
+   * @notice                Changes the endianness of a byte array
+   * @dev                   Returns a new, backwards, byte array
+   * @param {Uint8Array}    uint8Arr The array to reverse
+   * @returns {Uint8Array}  The reversed array
+   */
+  reverseEndianness: (uint8Arr) => {
+    let newArr = utils.safeSlice(uint8Arr);
+    return new Uint8Array(newArr.reverse());
+  },
+
+  /**
+   * @notice                Converts big-endian array to a uint
+   * @dev                   Traverses the byte array and sums the bytes
+   * @param {Uint8Array}    uint8Arr The big-endian array-encoded integer
+   * @returns {BigInt}      The integer representation
+   */
+  bytesToUint: (uint8Arr) => {
+    let total = BigInt(0);
+    for (let i = 0; i < uint8Arr.length; i++) {
+      total += BigInt(uint8Arr[i]) << (BigInt(uint8Arr.length - i - 1) * BigInt(8));
+    }
+    return total;
+  },
+
+  /**
+   * @notice                 Get the last _num bytes from a byte array
+   * @dev                    The byte array to slice
+   * @param {Uint8Array}     uint8Arr The big-endian array-encoded integer
+   * @returns {BigInt}       The integer representation
+   */
+  lastBytes: (arr, num) => {
+    if (num >= arr.length) {
+      throw new Error('Underflow during subtraction.');
+    }
+    return utils.safeSlice(arr, arr.length - num);
+  },
+
+  /**
+   * @notice                Implements bitcoin's hash160 (rmd160(sha2()))
+   * @param {Uint8Array}    preImage The pre-image
+   * @returns {Uint8Array}  The digest
+   */
+  hash160: (preImage) => {
+    return utils.ripemd160(utils.sha256(preImage));
+  },
+
+/**
+ * @notice                  Implements bitcoin's hash256 (double sha2)
+ * @param {Uint8Array}      preImage The pre-image
+ * @returns {Uint8Array}    The digest
+ */
+  hash256: (b) => {
+    return utils.sha256(utils.sha256(b));
+  },
+
+  /* ************ */
+  /* Legacy Input */
+  /* ************ */
+
+  /**
+   * @notice                Extracts the nth input from the vin (0-indexed)
+   * @dev                   Iterates over the vin. If you need to extract several, write a custom function
+   * @param {Uint8Array}    vinArr The vin as a tightly-packed uint8array
+   * @param {index}         index The 0-indexed location of the input to extract
+   * @returns {Uint8Array}  The input as a u8a
+   */
+  extractInputAtIndex: (vinArr, index) => {
+    let len = 0;
+    let remaining = 0;
+    let offset = 1n;
+
+    for (let i = 0; i <= index; i++) {
+      remaining = utils.safeSlice(vinArr, offset, vinArr.length - 1);
```

Comment:
Is this `- 1` supposed to be here?  Doesn't this make `remaining` always miss 1 byte off the end of the array?

Reply:
fixed